src/distribution/robust_soliton.rs

```diff
@@ -0,0 +1,345 @@
+        if max < 1 {
+            Err(StatsError::BadParams)
+        } else {
+            let pmf_table: Vec<f64> = Vec::with_capacity(max as usize);
+            let mut rs = RobustSoliton {
+                min: 1,
+                max,
+                spike: None,
+                cumulative_probability_table: pmf_table,
+                ripple: match heuristic {
+                    false => Ripple::Fixed(ripple),
+                    true => Ripple::Heuristic(ripple),
+                },
+                fail_probability,
+            };
+
+            let mut cumulative_probability = 0.0;
+            for i in 1..(max + 1) {
+                cumulative_probability += rs.soliton(i);
+                rs.cumulative_probability_table.push(cumulative_probability);
+            }
+            Ok(rs)
+        }
+    }
+
+    pub fn query_table<R: Rng + ?Sized>(&self, r: &mut R) -> Result<i64> {
+        let n = self.sample(r);
+
+        for i in 1..(self.max+1) {
+            if n < self.cumulative_probability_table[i as usize] {
+                return Ok(i);
+            }
+        }
+        Err(StatsError::ContainerExpectedSum("Elements in probability table expected to sum to 1 but didn't! Limit is", self.max as f64))
+    }
+
+    pub fn query_table_int<R: Rng + ?Sized>(&self, r: &mut R) -> Result<usize> {
+        let n = r.gen_range(self.min as usize, self.max as usize);
+        Ok(n)
+    }
+
+}
+
+#[derive(PartialEq, Clone, Debug)]
+enum Ripple {
+    Fixed(f64),
+    Heuristic(f64)
+}
+
+impl Ripple {
+    fn size(&self, max: i64, fail_probability: f64) -> f64 {
+        match self {
+            &Ripple::Fixed(n) => n,
+            &Ripple::Heuristic(n) => {
+                n * (max as f64 / fail_probability).ln() * (max as f64).sqrt()
+            }
+        }
+    }
+}
+
+impl Distribution<f64> for RobustSoliton {
+    fn sample<R: Rng + ?Sized>(&self, r: &mut R) -> f64 {
+        r.gen_range(0, 1) as f64
+    }
+}
+
+impl Min<i64> for RobustSoliton {
+    /// Returns the minimum value in the domain of the robust soliton
+    /// distribution
+    ///
+    /// # Remarks
+    ///
+    /// This is the same value as the minimum passed into the constructor
+    fn min(&self) -> i64 {
+        self.min
+    }
+}
+
+impl Max<i64> for RobustSoliton {
+    /// Returns the maximum value in the domain of the robust soliton
+    /// distribution
+    ///
+    /// # Remarks
+    ///
+    /// This is the same value as the maximum passed into the constructor
+    fn max(&self) -> i64 {
+        self.max
+    }
+}
+
+impl Mean<f64> for RobustSoliton {
+    /// Returns the mean of the robust soliton distribution
+    ///
+    /// # Formula
+    ///
+    /// ```ignore
+    /// sum(Xn)/n
+    /// ```
+    fn mean(&self) -> f64 {
+        let sum: f64 = Iterator::sum(self.cumulative_probability_table.iter());
+        let mean = sum / self.cumulative_probability_table.len() as f64;
+        mean
```

Comment:
This would be clearer as `.iter().sum()`

Can you explain how this calculation works? As far as I know, there are two equivalent formulas for calculating the mean: Sum(n*p(n)) or when using cumulative probability Sum(P(X>n)) = Sum(1-P(X<=n)) = Sum(1-CPF(n))
I do not see either of them being used here.